ui/src/app/views/workspace-share/component.ts

```diff
@@ -61,32 +69,72 @@ export class WorkspaceShareComponent implements OnInit {
     }
   }
 
+  convertToEmail(username: string): string {
+    return username + '@fake-research-aou.org';
+  }
+
+  inputChange(): void {
+    this.userNotFound = false;
+  }
+
   addCollaborator(): void {
-    this.workspace.userRoles.push({email: this.toShare, role: this.accessLevel});
-    this.workspacesService.shareWorkspace(
-      this.workspace.namespace, this.workspace.id, {
-        workspaceEtag: this.workspace.etag,
-        items: this.workspace.userRoles
-      }).subscribe((resp: ShareWorkspaceResponse) => {
-        this.workspace.etag = resp.workspaceEtag;
+    if (this.inputNotInitialized) {
+      this.inputNotInitialized = false;
+      this.input.nativeElement.addEventListener('keydown', () => {
+        this.userNotFound = false;
       });
+    }
+    this.usersLoading = true;
+    this.updateList = Array.from(this.workspace.userRoles);
+    this.updateList.push({email: this.convertToEmail(this.toShare),
+        role: this.accessLevel});
+
+    this.errorHandlingService.retryApi(
+      this.workspacesService.shareWorkspace(
+        this.workspace.namespace,
+        this.workspace.id, {
+          workspaceEtag: this.workspace.etag,
+          items: this.updateList})).subscribe(
+      (resp: ShareWorkspaceResponse) => {
+        this.workspace.etag = resp.workspaceEtag;
+        this.usersLoading = false;
+        this.workspace.userRoles = this.updateList;
+        this.toShare = '';
+        this.input.nativeElement.focus();
+      },
+      (error) => {
+        if (error.status === 400) {
```

Comment:
Not found is 400 and not a 404?

Reply:
Will update on API side (currently throwing a badRequestException?). I'm not actually sure if this should be considered a badRequest or a notFound.

Reply:
This should be a 404. 400 should be reserved for actually malformed http.

Reply:
I tend to use 404 and 405 to refer specifically to the URL and method, respectively. Otherwise, I'm not sure how to distinguish between, "your request is semantically invalid" and "you fat-fingered the URL." That leaves 400 to be fairly generic. For example, calling myservice/create-user with the payload `{"totallyInvalid": 7}` is a bad request because the payload doesn't have the required information. I would treat a request to /myservice/user/1234 as a 404 if that user doesn't exist because it is a URL. In this case, I'd argue that the request is semantically invalid because you're putting invalid data into the payload (a non-existant account), so I'd go with 400.

To put it another way, how would you distinguish between the following?
```
/api/v1/workspaces/my-namespace/my-workspace/share
{"email": "totally-invalid@researchallofus.org"}
```
and
```
/api/v1/workspaces/my-namespace/my-workspace/sh4re
{"email": "valid-user@researchallofus.org"}
```

Reply:
having taken a closer look at the [actual api method](https://github.com/all-of-us/workbench/blob/5ad30f86b7e5f7d53aa4b94656d4260edd2a8236/api/src/main/resources/workbench.yaml#L421) I'm going to agree with @dmohs here; if a user isn't found then in this case that indicates bad POST data (a 400).